portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,61 +14,76 @@
+  /** Header containing the total number of comments stored */
+  private static final String TOTAL_NUMBER_HEADER = "num-comments";
+  /** Query string that indicates whether to sort ascending or descending */
+  private static final String SORT_ASCENDING_QUERY = "sort-ascending";
+  /** Parameter which contains the ID's of the comments to delete */
+  private static final String DELETE_PARAMETER = "delete";
+
+  private static CommentPersistHelper comments;
 
   @Override
   public void init() {
-    // Retrieve stored comments
-    Query query = new Query("Comment").addSort(Comment.COMMENT_TIMESTAMP, SortDirection.DESCENDING);
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    PreparedQuery results = datastore.prepare(query);
-
-    results.asList(FetchOptions.Builder.withDefaults()).forEach(entity -> {
-      comments.add(Comment.fromEntity(entity));
-    });
+    comments = new CommentPersistHelper();
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Gson gson = new Gson();
+    int numComments = DEFAULT_COMMENT_NUM;
+    boolean sortAscending = true;
+
+    if(request.getParameter(NUM_COMMENTS_QUERY) != null) {
+      numComments = Integer.parseInt(request.getParameter(NUM_COMMENTS_QUERY));
+      // Revert to default if user specified an invalid number
+      if(numComments <= 0) {
+        numComments = DEFAULT_COMMENT_NUM;
+      }
+    }
+    if(request.getParameter(SORT_ASCENDING_QUERY) != null) {
+      sortAscending = Boolean.parseBoolean(request.getParameter(SORT_ASCENDING_QUERY));
+    }
+
     response.setContentType("application/json;");
-    response.getWriter().println(gson.toJson(comments));
+    response.getWriter().println(comments.stringifyComments(numComments, sortAscending));
+    // Send the total number of comments
+    response.addIntHeader(TOTAL_NUMBER_HEADER, comments.getNumComments());
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Create and add the new comment
-    Comment comment = Comment.fromHttpRequest(request);
-    comments.add(comment);
-
-    // Store the comment so it persists
-    Entity entity = comment.toEntity();
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    datastore.put(entity);
+    comments.addComment(request);
     
     // Redirect back to the home page.
     response.sendRedirect("/index.html");
   }
+
+  @Override
+  public void doDelete(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String[] toDelete = request.getParameterValues(DELETE_PARAMETER);
+    Arrays.stream(toDelete).forEach(idStr -> {
+      try {
+        int id = Integer.parseInt(idStr);
+        comments.deleteComment(id);
+      } catch(NumberFormatException e) {
+        // Ignore, as should never happen
```

Comment:
when this does happen, instead you should probably throw an IllegalArgumentException